src/library/serato/seratofeature.cpp

```diff
@@ -0,0 +1,1214 @@
+                query.bindValue(":bitrate", track.bitrate);
+                query.bindValue(":samplerate", track.samplerate);
+                query.bindValue(":bpm", track.bpm);
+                query.bindValue(":key", track.key);
+                query.bindValue(":location", location);
+                query.bindValue(":bpm_lock", track.beatgridlocked);
+                query.bindValue(":datetime_added", track.datetimeadded);
+                query.bindValue(":label", track.label);
+                query.bindValue(":serato_db", databaseDir.path());
+
+                if (!query.exec()) {
+                    LOG_FAILED_QUERY(query);
+                } else {
+                    int trackId = query.lastInsertId().toInt();
+                    insertTrackIntoPlaylist(database, playlistId, trackId, trackCount);
+                    trackIdMap.insert(track.location, trackId);
+                    trackCount++;
+                }
+                break;
+            }
+            break;
+        }
+        default: {
+            QString fieldName = QString(headerData.mid(0, sizeof(quint32)));
+            qDebug() << "Ignoring unknown field "
+                     << fieldName
+                     << " ("
+                     << fieldSize
+                     << " bytes) in database "
+                     << databaseFilePath
+                     << ".";
+        }
+        }
+
+        headerData = databaseFile.read(kHeaderSize);
+    }
+
+    if (headerData.length() != 0) {
+        qWarning() << "Found "
+                   << headerData.length()
+                   << " extra bytes at end of Serato database file "
+                   << databaseFilePath
+                   << ".";
+    }
+
+    // Parse Crates
+    QDir crateDir = QDir(databaseDir);
+    if (crateDir.cd(kCrateDirectory)) {
+        QStringList filters;
+        filters << kCrateFilter;
+        foreach (const QString& entry, crateDir.entryList(filters)) {
+            QString crateFilePath = crateDir.filePath(entry);
+            QString crateName = parseCrate(
+                    database,
+                    databaseDir.path(),
+                    crateFilePath,
+                    trackIdMap);
+            if (!crateName.isEmpty()) {
+                QList<QVariant> data;
+                data << QVariant(crateFilePath)
+                     << QVariant(true);
+                TreeItem* crateItem = databaseItem->appendChild(crateName, data);
+                crateItem->setIcon(QIcon(":/images/library/ic_library_crates.svg"));
+            }
+        }
+    } else {
+        qWarning() << "Failed to open crate directory: "
+                   << databaseDir.filePath(kCrateDirectory);
+    }
+
+    // TODO: Parse Smart Crates
+
+    transaction.commit();
+
+    return databaseFilePath;
+}
+
+// This function is executed in a separate thread other than the main thread
+QList<TreeItem*> findSeratoDatabases(SeratoFeature* seratoFeature) {
```

Comment:
Passing a pointer to a QObject that lives in the main thread to a worker thread is considered unsafe. PR #2502 fixes this unnecessary dependency.

Reply:
I noticed this flaw during the review, not your fault ;)